examples/service_worker/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+                serde_json::to_string_pretty(&push_subscription).unwrap()
+            );
+            model.notifications_granted = true;
+            model.push_subscription = Some(push_subscription);
+            orders.send_msg(Msg::StatusUpdate(
+                ServiceWorkerStatus::SubscriptionRetrieved,
+            ));
+        }
+        Msg::SendMessage => {
+            web_sys::Notification::new("Hello from Seed service worker!")
+                .expect("Couldn't send notification.");
+        }
+        Msg::SetServiceWorker(sw_reg) => {
+            model.sw_reg = Some(sw_reg);
+        }
+    }
+}
+
+// ------ ------
+//     View
+// ------ ------
+
+fn view(model: &Model) -> Node<Msg> {
+    div![
+        h1!["Seed - Service Worker Demo"],
+        p!["When the page first loads, the service worker will cache all assets, but permissions for notifications must be granted."],
+        ol![
+            li![
+                "Click the ",
+                b!["Request Push Subscription"],
+                " button and when prompted, select ",
+                b!["Allow Notifications."],
+            ],
+            li!["The subscription information will be printed to the page and notifications will now be granted."],
+            li![
+                "Click the ",
+                b!["Send Message"],
+                " button to see a browser notification coming from service worker.",
+            ],
+        ],
+        button![ev(Ev::Click, |_| Msg::RequestNotificationPermission), "Request Push Subscription"],
+        button![
+            ev(Ev::Click, |_| Msg::SendMessage),
+            IF!(not(model.notifications_granted) => attrs!{At::Disabled => true}),
+            "Send Message"
+        ],
+        IF!(model.error.is_some() => {
+            p![
+                attrs!{At::Style => "color:red"},
+                model.error.clone().unwrap()
+            ]
+        }),
+        IF!(model.message.is_some() => {
+            p![
+                model.message.clone().unwrap()
+            ]
+        }),
+        h2!["Push Subscription"],
+        code![serde_json::to_string_pretty(&model.push_subscription).unwrap()],
+        br![],
+        br![],
+        img![attrs! {
+            At::Src => "images/important-notes.png"
+        },],
+        br![],
+        a![
+            attrs! {
+                At::Href => "https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers"
+            },
+            "- Using Service Workers (Service Worker Api)"
+        ]
+    ]
+}
+
+// ------ ------
+//     Start
+// ------ ------
+
+#[wasm_bindgen(start)]
+pub fn start() {
+    let app = App::start("app", init, update, view);
+
+    spawn_local(async move {
+        match register_service_worker(app.clone()).await {
+            Ok(x) => x,
+            Err(e) => {
+                app.update(Msg::FailedToRegisterServiceWorker(e));
+            }
+        }
+    });
```

Comment:
Is there a reason why it can't be placed in our `init` function?
If it can be moved to `init`, please move also `register_service_worker` just below the `init` function.

Reply:
It's funny you mention that as I started out by putting it in the init function. I don't think I'll run into any issues but if I do I'll let you know :)

Reply:
I'm having some trouble with this so maybe you can point me in the right direction. I've tried a couple of things:
1. Pass orders to the function so I can use it to send messages:
```
fn init(_: Url, orders: &mut impl Orders<Msg>) -> Model {
    spawn_local(async move {
        //let app: App<Msg, Model, Node<Msg>> = app.clone_into(target)
        match register_service_worker(orders).await {
            Ok(x) => x,
            Err(e) => {
                //App::update(Msg::FailedToRegisterServiceWorker(e));
            }
        }
    });

    Model::default()
}
```
Produces:
```
error[E0621]: explicit lifetime required in the type of `orders`
  --> examples/service_worker/src/lib.rs:24:5
   |
23 | fn init(_: Url, orders: &mut impl Orders<Msg>) -> Model {
   |                         --------------------- help: add explicit lifetime `'static` to the type of `orders`: `&'static mut impl Orders<Msg>`
24 |     spawn_local(async move {
   |     ^^^^^^^^^^^ lifetime `'static` required
```
However, it doesn't look like I can make orders static.

2. Get the app from orders and attempt to pass it in:
```
fn init(_: Url, orders: &mut impl Orders<Msg>) -> Model {
    spawn_local(async move {
        //let app: App<Msg, Model, Node<Msg>> = app.clone_into(target)
        match register_service_worker(orders.clone_app()).await {
            Ok(x) => x,
            Err(e) => {
                //App::update(Msg::FailedToRegisterServiceWorker(e));
            }
        }
    });

    Model::default()
}
```
Yields:
```
error[E0308]: mismatched types
  --> examples/service_worker/src/lib.rs:26:39
   |
26 |         match register_service_worker(orders.clone_app()).await {
   |                                       ^^^^^^^^^^^^^^^^^^ expected enum `Msg`, found associated type
   |
   = note: expected struct `seed::App<Msg, Model, seed::virtual_dom::Node<Msg>>`
              found struct `seed::App<<impl Orders<Msg> as seed::app::Orders<Msg>>::AppMs, <impl Orders<Msg> as seed::app::Orders<Msg>>::Mdl, <impl Orders<Msg> as seed::app::Orders<Msg>>::INodes>`
   = help: consider constraining the associated type `<impl Orders<Msg> as seed::app::Orders<Msg>>::AppMs` to `Msg`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
```

I'm probably missing something basic here. What' the best way to pass either the orders or the app?

Reply:
<details>
<summary>One of the options how to invoke `register_service_worker` in `init`</summary>

```rust
fn init(_: Url, orders: &mut impl Orders<Msg>) -> Model {
    let msg_sender = orders.msg_sender();

    orders.perform_cmd(async move {
        match register_service_worker(Rc::clone(&msg_sender)).await {
            Ok(x) => x,
            Err(e) => {
                msg_sender(Some(Msg::FailedToRegisterServiceWorker(e)));
            }
        }
    });

    Model::default()
}

async fn register_service_worker(
    msg_sender: Rc<dyn Fn(Option<Msg>)>,
) -> Result<(), ServiceWorkerError> {
    let window = web_sys::window().ok_or(ServiceWorkerError::GetWindow)?;
    let sw_container = window.navigator().service_worker();

    let p = sw_container.register("service-worker.js");
    let reg = wasm_bindgen_futures::JsFuture::from(p)
        .await
        .map_err(ServiceWorkerError::Registration)?;

    let sw_reg: web_sys::ServiceWorkerRegistration = reg.into();

    msg_sender(Some(Msg::SetServiceWorker(sw_reg.clone())));
    let sw: Option<web_sys::ServiceWorker> = if let Some(x) = sw_reg.installing() {
        Some(x)
    } else if let Some(x) = sw_reg.waiting() {
        Some(x)
    } else if let Some(x) = sw_reg.active() {
        Some(x)
    } else {
        None
    };

    if let Some(sw) = sw {
        if sw.state() == web_sys::ServiceWorkerState::Activated {
            msg_sender(Some(Msg::ServiceWorkerActivated));
            msg_sender(Some(Msg::StatusUpdate(ServiceWorkerStatus::Activated)));
        } else {
            msg_sender(Some(Msg::StatusUpdate(ServiceWorkerStatus::Waiting)));
            let c = Closure::wrap(Box::new(move |e: web_sys::Event| {
                let target: web_sys::EventTarget = e.target().expect("Couldn't get target");
                let sw_js: JsValue = target.into();
                let sw: web_sys::ServiceWorker = sw_js.into();
                let state: web_sys::ServiceWorkerState = sw.state();

                msg_sender(Some(Msg::StatusUpdate(ServiceWorkerStatus::StateChange(state))));

                if state == web_sys::ServiceWorkerState::Activated {
                    msg_sender(Some(Msg::ServiceWorkerActivated));
                }
            }) as Box<dyn Fn(web_sys::Event)>);

            sw.add_event_listener_with_callback("statechange", c.as_ref().unchecked_ref())
                .map_err(|v| {
                    sw.remove_event_listener_with_callback(
                        "statechange",
                        c.as_ref().unchecked_ref(),
                    )
                    .expect("Couldn't remove state change event listener.");
                    ServiceWorkerError::StateChangeListener(v)
                })?;

            c.forget();
        }
    } else {
        return Err(ServiceWorkerError::InvalidState);
    }

    Ok(())
}
```

</details>
Notice changed `spawn_local` to `orders.perform_cmd` and `app.update` to `msg_sender`.

Reply:
Fantastic! This is exactly what I needed.